### PR TITLE
[G-API] GStreamingExecutor abstract interface

### PR DESCRIPTION
**Pull request by OrestChura**
This pull request separates `GStreamingExecutor` into interface and `GTBBStreamingExecutor` which is a TBB implementation for the streaming mode executor.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/gapi/src/executor/gtbbstreamingexecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,1728 @@
...
+            video_emitters.emplace_back(emit_nh);
+#else
+            util::throw_error(std::logic_error("Video is not supported in the "
+                                               "standalone mode"));
+#endif
+            break;
+        default:
+            // Create a constant emitter.
+            // Produces always the same ("constant") value when pulled.
+            emitter.reset(new ConstEmitter{emit_arg});
```
Comment (sivanov-work) [inline modules/gapi/src/executor/gtbbstreamingexecutor.cpp:1462]: GAPI_LOG_INFO("create ConstEmitter" );

File: modules/gapi/src/executor/gtbbstreamingexecutor.hpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    virtual void push(const Cmd &cmd) override { m_v.push(cmd); }
+    virtual void pop(Cmd &cmd)        override { m_v.pop(cmd);  }
+    virtual bool try_pop(Cmd &cmd)    override { return m_v.try_pop(cmd); }
+    virtual void clear()              override { m_v.clear(); }
+};
+} // namespace stream
+
+class GTBBStreamingExecutor final : public GStreamingExecutor {
+protected:
+    const bool m_desync;
```
Comment (sivanov-work) [inline modules/gapi/src/executor/gtbbstreamingexecutor.hpp:94]: I suggest to implement TWO different `GTBBStreamingExecutor`:
`GTBB_SYNC_StreamingExecutor` ---> `GTBBStreamingExecutor` <-- `GTBB_DESYNC_StreamingExecutor`

Do do not stick flag-based logic through single and huge GTBBStreamingExecutor

**Comment by sivanov-work** [comment]
I suggest that we separated utils-like  `queue`s , `reader` and so on into different files. Because it is hard to investigate these. Also they could be reused in different components
Also i suggest that we should move out `const desync` flag and apply further division of TBB Executor on multiple descendants: for sync & desync cases.

File: modules/gapi/src/compiler/gstreaming.cpp [hunk 3]
```diff
@@ -72,12 +73,16 @@ bool cv::GStreamingCompiled::Priv::pull(cv::GRunArgsP &&outs)
 
 bool cv::GStreamingCompiled::Priv::pull(cv::GOptRunArgsP &&outs)
 {
-    return m_exec->pull(std::move(outs));
+    auto* exec = dynamic_cast<cv::gimpl::GTBBStreamingExecutor*>(m_exec.get());
```
Comment (smirnov-alexey) [inline modules/gapi/src/compiler/gstreaming.cpp:76]: Why do we need those checks in each `pull`?

File: modules/gapi/src/executor/gstreamingexecutor.hpp [hunk 4]
```diff
@@ -138,75 +60,22 @@ class GStreamingExecutor final
...
-        std::vector<stream::Q*>  queues;
-        std::vector<int> mapping;
-    };
-    std::unordered_map<int, CollectorThreadInfo> m_collector_map;
-
-
-    void wait_shutdown();
-
-    cv::GTypesInfo out_info;
+    void virtual start_impl() = 0;
```
Comment (smirnov-alexey) [inline modules/gapi/src/executor/gstreamingexecutor.hpp:66]: Could you, please, explain why `_impl`s are `virtual` and not the default functions?

File: modules/gapi/src/executor/gtbbstreamingexecutor.hpp (hunk 2 above)
Comment (smirnov-alexey) [reply modules/gapi/src/executor/gtbbstreamingexecutor.hpp:94]: @sivanov-work wouldn't it involve a huge chunk of code being copy-pasted?

File: modules/gapi/src/executor/gtbbstreamingexecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,1728 @@
...
+
+    // FIXME: ADE metadata requires types to be copiable
+    std::shared_ptr<Q> q;
+};
+
+struct DesyncSpecialCase {
+    static const char *name() { return "DesyncSpecialCase"; }
+};
+
+std::vector<Q*> reader_queues(      ade::Graph &g,
```
Comment (smirnov-alexey) [inline modules/gapi/src/executor/gtbbstreamingexecutor.cpp:88]: Alignment is broken here and below

File: modules/gapi/src/executor/gtbbstreamingexecutor.cpp [hunk 6]
```diff
@@ -0,0 +1,1728 @@
...
+
+    for (auto && it : ade::util::zip(ade::util::toRange(outputs),
+                                     ade::util::toRange(r.args),
+                                     ade::util::toRange(r.flags))) {
+        auto &out_obj  = std::get<0>(it);
+        auto &res_obj  = std::get<1>(it);
+        bool available = std::get<2>(it);
+
+        using T = cv::GOptRunArgP;
+#define HANDLE_CASE(Type)                                               \
```
Comment (smirnov-alexey) [inline modules/gapi/src/executor/gtbbstreamingexecutor.cpp:123]: Please, align the `\`

File: modules/gapi/src/executor/gtbbstreamingexecutor.cpp [hunk 7]
```diff
@@ -0,0 +1,1728 @@
...
+            HANDLE_CASE(cv::Scalar);     break;
+            HANDLE_CASE(cv::RMat);       break;
+            HANDLE_CASE(cv::MediaFrame); break;
+
+        case T::index_of<O<cv::Mat>*>(): {
+            // Mat: special handling.
+            auto &mat_opt = *cv::util::get<O<cv::Mat>*>(out_obj);
+            if (available) {
+                auto q_map = cv::util::get<cv::RMat>(res_obj).access(cv::RMat::Access::R);
+                // FIXME: Copy! Maybe we could do some optimization for this case!
```
Comment (smirnov-alexey) [inline modules/gapi/src/executor/gtbbstreamingexecutor.cpp:143]: Shouldn't we be concerned about this copy? Is there a task to fix it later if needed?

File: modules/gapi/src/executor/gstreamingexecutor.cpp [hunk 8]
```diff
@@ -1292,658 +70,77 @@ cv::gimpl::GStreamingExecutor::GStreamingExecutor(std::unique_ptr<ade::Graph> &&
...
 
-cv::gimpl::GStreamingExecutor::~GStreamingExecutor()
-{
+cv::gimpl::GStreamingExecutor::~GStreamingExecutor() {
     // FIXME: this is a temporary try-catch exception hadling.
-    // Need to eliminate throwings from stop()
+    // Need to eliminate throwings from stop(): make stop_impl() & wait_shutdown() no_throw
     try {
-        if (state == State::READY || state == State::RUNNING)
+        if (state == State::READY || state == State::RUNNING) {
```
Comment (smirnov-alexey) [inline modules/gapi/src/executor/gstreamingexecutor.cpp:80]: Shouldn't we make `state` an `atomic`/guard it? I guess the question is not related to this PR but still

Comment (smirnov-alexey) [reply modules/gapi/src/executor/gstreamingexecutor.cpp:80]: Please, ignore if it's not a critical section

**Pull request closed without merging**